### ISSUE TITLE
Deprecated config options are not processed, as if they were Invalid

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#18969]
Commit 5fd30578d74ba94ca4a132924c170cd41ae590dc introduced a `value_status::Deprecated` that produces a log warning when the option is used in the configuration file.
See https://github.com/scylladb/scylladb/blob/8a77a74d0ec6b7e988c7a073e82226889c0bc8ba/utils/config_file.cc#L317-L327

But the way it was introduced, `Deprecated` is treated exactly like `Invalid`, and the `continue;` statement in the switch cuts processing short, before getting to https://github.com/scylladb/scylladb/blob/8a77a74d0ec6b7e988c7a073e82226889c0bc8ba/utils/config_file.cc#L333

causing the option to be ignored.

This contradicts typical deprecation semantics that allow users to continue using the interface while its deprecated, under the warning that the support for it is about to be dropped in the future (not at the time the interface is declared as deprecated).

**Comment by xemul** [comment]
Deprecated was introduced to solve backward compatibility problem. The option is marked as such should disappear from help text, but not to fail parser if met

**Comment by bhalevy** [comment]
We need a test to cement our collective expectations for the expected semantics...

**Comment by bhalevy** [comment]
I see deprecated options, like `-override-decommission`, in `scylla --help`

**Comment by ptrsmrn** [comment]
> I see deprecated options, like `--override-decommission`, in `scylla --help`

Yes, but it appears in the deprecated options section:
```

**Comment by ptrsmrn** [comment]
> Deprecated was introduced to solve backward compatibility problem. The option is marked as such should disappear from help text, but not to fail parser if met

So it doesn't fail the parser and IS present in the `--help` (https://github.com/scylladb/scylladb/issues/18969#issuecomment-2138204693)

**Comment by ptrsmrn** [comment]
> [...] causing the option to be ignored.
> 
> This contradicts typical deprecation semantics that allow users to continue using the interface while its deprecated, under the warning that the support for it is about to be dropped in the future (not at the time the interface is declared as deprecated).

In the light of your reasoning, I agree that the naming for this `Deprecated` enumerator was unfortunate - I agree that usually _deprecated_ means that the parameter is handled as usual, but is about to be stopped handled at all, whereas here we stopped stopped assigning it a value. But this was a choice we agreed on in: https://docs.google.com/document/d/18urPG7qeb7z7WPpMYI2V_lCOkM5YGKsEU78SDJmt8bM/edit?usp=sharing:
> This new Deprecated marker would make the config option be displayed in scylla’s --help, but not parsed

I'm also unsure if we can freely change how config options are handled back and forth all the time - we were doing it and still continue to.

**Comment by bhalevy** [comment]
> > [...] causing the option to be ignored.
> > This contradicts typical deprecation semantics that allow users to continue using the interface while its deprecated, under the warning that the support for it is about to be dropped in the future (not at the time the interface is declared as deprecated).
> 
> In the light of your reasoning, I agree that the naming for this `Deprecated` enumerator was unfortunate - I agree that usually _deprecated_ means that the parameter is handled as usual, but is about to be stopped handled at all, whereas here we stopped stopped assigning it a value. But this was a choice we agreed on in: https://docs.google.com/document/d/18urPG7qeb7z7WPpMYI2V_lCOkM5YGKsEU78SDJmt8bM/edit?usp=sharing:
> 
> > This new Deprecated marker would make the config option be displayed in scylla’s --help, but not parsed
> 
> I'm also unsure if we can freely change how config options are handled back and forth all the time - we were doing it and still continue to.

If we made a mistake, we need to fix it, no matter how many times we got it wrong.

https://en.m.wikipedia.org/wiki/Deprecation
```
Deprecation is the discouragement of use of something human-made, such as a [term](https://en.m.wikipedia.org/wiki/Terminology), feature, [design](https://en.m.wikipedia.org/wiki/Design), or practice. Typically something is deprecated because it is claimed to be inferior compared to other options available.

Something may be deprecated when it cannot be controlled,[[clarification needed](https://en.m.wikipedia.org/wiki/Wikipedia:Please_clarify)] such as a term. Even when it can be controlled, something may be deprecated even when it might be useful – for example, to ensure compatibility – and it may be removed or discontinued at some time after being deprecated.
```

I don't see anything in the above that says that deprecated options are ignored.  That's extremely misleading.

If we want to keep some old options around for backward compatibility and now they have no effect, they need to be clearly described as "Unused" or "Ignored".  Why call them "deprecated"?

**Comment by dani-tweig** [comment]
@bhalevy is this issue still relevant?
@mykaul , as P4, can we move it to the backlog milestone?

**Comment by bhalevy** [comment]
> [@bhalevy](https://github.com/bhalevy) is this issue still relevant?

Yes.  Nothing has changed in this AFAIK and our config options depreciation semantics are still weird.

> [@mykaul](https://github.com/mykaul) , as P4, can we move it to the backlog milestone?

I have no objection
